accessible/tree.service.js

```diff
@@ -0,0 +1,325 @@
+          }
+      }
+      this.previousKey_ = e.keyCode;
+    },
+    getFirstChild: function(element) {
+      if (element == null) {
+        return element;
+      } else {
+        var childList = element.children;
+        for (var i = 0; i < childList.length; i++) {
+          if (childList[i].tagName == 'LI') {
+            return childList[i];
+          } else {
+            var potentialElement = this.getFirstChild(childList[i]);
+            if (potentialElement) {
+              return potentialElement;
+            }
+          }
+        }
+        return null;
+      }
+    },
+    getNextSibling: function(element) {
+      if (element.nextElementSibling) {
+        // if there is a sibling, find the list element child of the sibling
+        var node = element.nextElementSibling;
+        if (node.tagName != 'LI') {
+          var listElems = node.getElementsByTagName('li');
+          // getElementsByTagName returns in DFS order
+          // therefore the first element is the first relevant list child
+          return listElems[0];
+        } else {
+          return element.nextElementSibling;
+        }
+      } else {
+        var parent = element.parentNode;
+        while (parent && parent.tagName != 'OL') {
+          if (parent.nextElementSibling) {
+            var node = parent.nextElementSibling;
+            if (node.tagName == 'LI') {
+              return node;
+            } else {
+              return this.getFirstChild(node);
+            }
+          } else {
+            parent = parent.parentNode;
+          }
+        }
+        return null;
+      }
+    },
+    getPreviousSibling: function(element) {
+      if (element.previousElementSibling) {
+        var sibling = element.previousElementSibling;
+        if (sibling.tagName == 'LI') {
+          return sibling;
+        } else {
+          return this.getLastChild(sibling);
+        }
+      } else {
+        var parent = element.parentNode;
+        while (parent != null) {
+          blocklyApp.debug && console.log('looping');
+          if (parent.tagName == 'OL') {
+            break;
+          }
+          if (parent.previousElementSibling) {
+            blocklyApp.debug && console.log('parent has a sibling!');
+            var node = parent.previousElementSibling;
+            if (node.tagName == 'LI') {
+              // the parent has a list sibling!
+              blocklyApp.debug && console.log('return the sibling of the parent!');
+              return node;
+            } else {
+              // find the last list element child of the sibling of the parent
+              return this.getLastChild(node);
+            }
+          } else {
+            parent = parent.parentNode;
+          }
+        }
+        return null;
+      }
+    },
+    getLastChild: function(element) {
+      if (element == null) {
+        blocklyApp.debug && console.log('no element');
+        return element;
+      } else {
```

Comment:
Why not:

listItems = element.getElementsByTagName('LI');
return listItems.length ? listItems[listItems.length - 1] : null;